### PR TITLE
Adapt it for ad9361 serial and network interface

### PR DESCRIPTION
**Pull request by cristiantpop**
Add support for new commands: READBUF, WRITEBUF
Add instance_id
Add non tinyiiod_read_nonblocking(), used to write buffer. Serial protocol sends a stream of data until it finishes.

### CODE REVIEW

**Comment by commodo** [comment]
this patch does too much in one go
maybe split the whitespace/astyle/style/various stuff into it's own commit;

it also needs a `Signed-off-by` tag

File: parser.c [hunk 1]
```diff
@@ -103,28 +108,39 @@ static int parse_open_string(struct tinyiiod *iiod, char *str)
 		return -EINVAL;
 
 	str = ptr + 1;
+	mask = strtoul(str, NULL, 16);
```
Comment (commodo) [inline parser.c:111]: the `strtoul` conversion should be it's own commit as well
though, i am not sure we should do it

i suspect the reason for this code [to be as it is] is to make it adaptable for really simple micro-controllers;
`strtoul` is a libc library that may not be present all the time;
maybe, `compat.c / compat.h` file could be added to adapt `strtoul` to systems where this isn't present

File: tinyiiod.c [hunk 2]
```diff
@@ -41,19 +45,27 @@ struct tinyiiod * tinyiiod_create(const char *xml,
 	return iiod;
 }
 
+/***************************************************************************//**
+ * @brief tinyiiod_destroy
+*******************************************************************************/
```
Comment (commodo) [inline tinyiiod.c:50]: these doc-strings also look like they should be their own commit

File: tinyiiod.c [hunk 3]
```diff
@@ -62,160 +74,219 @@ int tinyiiod_read_command(struct tinyiiod *iiod)
...
-	ssize_t ret;
-
-	if (channel)
-		ret = iiod->ops->ch_read_attr(device, channel,
-				ch_out, attr, buf, sizeof(buf));
-	else
-		ret = iiod->ops->read_attr(device, attr,
-				buf, sizeof(buf), debug);
-
+	char buf[0x1000];
```
Comment (commodo) [inline tinyiiod.c:173]: this is too much on some systems, where stack memory is limited

File: tinyiiod-private.h [hunk 4]
```diff
@@ -21,30 +21,30 @@
 #include "tinyiiod.h"
 
 char tinyiiod_read_char(struct tinyiiod *iiod);
-void tinyiiod_read(struct tinyiiod *iiod, char *buf, size_t len);
-ssize_t tinyiiod_read_line(struct tinyiiod *iiod, char *buf, size_t len);
```
Comment (commodo) [inline tinyiiod-private.h:25]: removing the `, size_t len);` argument is a bad idea;
the `tinyiiod_read_line()` function won't know how large the `buf` memory is

File: parser.c [hunk 5]
```diff
@@ -142,49 +158,52 @@ static int parse_readbuf_string(struct tinyiiod *iiod, char *str)
...
 
-	if (!strncmp(str, "VERSION", sizeof("VERSION"))) {
+	if (!strncmp(str, "VERSION", sizeof("VERSION") - 1)) {
 		char buf[32];
 
-		snprintf(buf, sizeof(buf), "%u.%u.%-7.7s\n",
-				TINYIIOD_VERSION_MAJOR,
-				TINYIIOD_VERSION_MINOR,
-				TINYIIOD_VERSION_GIT);
+		snprintf(buf, sizeof(buf), "%d.%d.%-7.7s\n",
```
Comment (commodo) [inline parser.c:176]: curios: why is this converted from `%u.%u` -> `%d.%d` ?

File: example.c [hunk 6]
```diff
@@ -48,25 +91,24 @@ static ssize_t read_attr(const char *device, const char *attr,
 }
 
 static ssize_t write_attr(const char *device, const char *attr,
-		const char *buf, size_t len, bool debug)
+			  const char *buf, size_t len, bool debug)
 {
 	return -ENOSYS;
 }
-
+int32_t scale = 0;
```
Comment (commodo) [inline example.c:98]: this `scale` variable could be made `static` and moved somewhere in the first part of the file, before the function definitions

File: parser.c [hunk 7]
```diff
@@ -60,7 +63,7 @@ static int parse_rw_string(struct tinyiiod *iiod, char *str, bool write)
 	}
 
 	ptr = strchr(str, ' ');
-	if (!!ptr ^ write)
+	if ((!!ptr )^ write)
```
Comment (commodo) [inline parser.c:66]: this should also be it's own commit ; and have a detailed explanation about it;
the reasoning is that it touches on library parts of libtinyiiod, and these should be explained when changing;
this is to avoid bad practices of changing things without any reasoning [even when the reason could be obvious]

File: parser.c [hunk 8]
```diff
@@ -142,49 +158,52 @@ static int parse_readbuf_string(struct tinyiiod *iiod, char *str)
 		return -EINVAL;
 
 	tinyiiod_do_readbuf(iiod, device, (size_t) bytes_count);
+
```
Comment (commodo) [inline parser.c:161]: this line change looks un-necessary;
while i agree that it may look better with this newline, it's a good idea to add a separate commit

File: parser.c [hunk 9]
```diff
@@ -142,49 +158,52 @@ static int parse_readbuf_string(struct tinyiiod *iiod, char *str)
...
+int32_t tinyiiod_parse_string(struct tinyiiod *iiod, char *str)
 {
-	while (*str == '\n' || *str == '\r')
-		str++;
-
 	if (str[0] == '\0')
 		return 0;
 
-	if (!strncmp(str, "VERSION", sizeof("VERSION"))) {
+	if (!strncmp(str, "VERSION", sizeof("VERSION") - 1)) {
```
Comment (commodo) [inline parser.c:173]: this looks like a fix commit;
typically, fix commits should be separate, and have an explanation of what they fix

File: example.c [hunk 10]
```diff
@@ -15,24 +15,67 @@
...
+		return -EINVAL;
+	else
+		return value;
+}
+
+static int32_t read_data(int32_t *instance_id, char *buf, size_t len)
+{
+	*instance_id = 0;
 	fread(buf, 1, len, stdin);
+	return 1;
```
Comment (mirceacaprioru) [inline example.c:55]: Usually the return is separated from the rest of the code with a blank line. Also return 1 in this case is considered a succes, are there any SUCCESS, FAIL defines somewhere in the network part ?

File: tinyiiod-private.h [hunk 11]
```diff
@@ -21,30 +21,30 @@
...
+int32_t tinyiiod_read_line(struct tinyiiod *iiod, char *buf);
 
 void tinyiiod_write_char(struct tinyiiod *iiod, char c);
 void tinyiiod_write(struct tinyiiod *iiod, const char *data, size_t len);
 void tinyiiod_write_string(struct tinyiiod *iiod, const char *str);
-void tinyiiod_write_value(struct tinyiiod *iiod, int value);
+void tinyiiod_write_value(struct tinyiiod *iiod, int32_t value);
```
Comment (commodo) [inline tinyiiod-private.h:29]: `int` is a more standard type that `int32_t` ; so, i would not change this ;
if we want to change this, we may need a `compat.h` layer to make sure that these types (`int32_t` & friends) exist; 

i agree that this code already has `ssize_t` types, which i'm also not sure how standard they are ;

in any case, if we want to fix/address int/int32_t/size_t types, we should do it in separate commits, and focus the discussion only on these

File: tinyiiod.c [hunk 12]
```diff
@@ -16,19 +16,23 @@
  */
 
 #include "tinyiiod-private.h"
-
```
Comment (commodo) [inline tinyiiod.c:19]: this removes a newline;

File: tinyiiod.c [hunk 13]
```diff
@@ -41,19 +45,27 @@ struct tinyiiod * tinyiiod_create(const char *xml,
...
 
-int tinyiiod_read_command(struct tinyiiod *iiod)
+/***************************************************************************//**
+ * @brief tinyiiod_read_command
+*******************************************************************************/
+int32_t tinyiiod_read_command(struct tinyiiod *iiod)
 {
-	char buf[128];
-	int ret;
+	char buf[128] = {0};
```
Comment (commodo) [inline tinyiiod.c:61]: initializing a 128 buffer size on the stack, is a bit costly for some micro-controllers;
it may be non-standard, since this construct may only be accepted by some compilers and not by others;
it's also un-needed; for buffers, you typically keep an offset & length of data where you look into/for

File: tinyiiod.h [hunk 14]
```diff
@@ -22,44 +22,55 @@
...
 #define _SSIZE_T_DEFINED
 #endif
 
 struct tinyiiod;
 
 struct tinyiiod_ops {
+	int32_t (*read)(int32_t *instance_id, char *buf, size_t len);
 	/* Read from the input stream */
-	void (*read)(char *buf, size_t len);
+	int32_t (*read_line)(int32_t *instance_id, char *buf);
```
Comment (commodo) [inline tinyiiod.h:37]: this looks like an API change
i am not sure how spread/used this code is;
but, if changing API, we may want to start a larger discussion about it;

in any case, these should be as part of separate commits, and have some more details about them

File: tinyiiod.h [hunk 15]
```diff
@@ -22,44 +22,55 @@
...
+	int32_t (*get_mask)(const char *device, uint32_t *mask);
 };
 
 struct tinyiiod * tinyiiod_create(const char *xml,
-		const struct tinyiiod_ops *ops);
+				  const struct tinyiiod_ops *ops);
 void tinyiiod_destroy(struct tinyiiod *iiod);
-int tinyiiod_read_command(struct tinyiiod *iiod);
+int32_t tinyiiod_read_command(struct tinyiiod *iiod);
+int32_t tinyiiod_do_writebuf(struct tinyiiod *iiod, const char *device,
```
Comment (commodo) [inline tinyiiod.h:72]: the way `tinyiiod_read_command()` and `tinyiiod_do_writebuf()`, are structured, they don't look like API-style functions ; 
maybe they belong in ` tinyiiod-private.h` ?

File: tinyiiod.c [hunk 16]
```diff
@@ -62,160 +74,219 @@ int tinyiiod_read_command(struct tinyiiod *iiod)
...
+ * @brief tinyiiod_do_write_attr
+*******************************************************************************/
 void tinyiiod_do_write_attr(struct tinyiiod *iiod, const char *device,
-		const char *channel, bool ch_out, const char *attr,
-		size_t bytes, bool debug)
+			    const char *channel, bool ch_out, const char *attr,
+			    size_t bytes, bool debug)
 {
-	char buf[128];
+	char buf[2048] = {0};
```
Comment (commodo) [inline tinyiiod.c:197]: this also looks a bit big for some stacks of some systems

File: tinyiiod.c [hunk 17]
```diff
@@ -62,160 +74,219 @@ int tinyiiod_read_command(struct tinyiiod *iiod)
...
 	tinyiiod_write_string(iiod, buf);
 }
 
+/***************************************************************************//**
+ * @brief tinyiiod_write_xml
+*******************************************************************************/
 void tinyiiod_write_xml(struct tinyiiod *iiod)
 {
 	size_t len = strlen(iiod->xml);
-
```
Comment (commodo) [inline tinyiiod.c:127]: this line removal looks un-necessary

File: tinyiiod.c [hunk 18]
```diff
@@ -62,160 +74,219 @@ int tinyiiod_read_command(struct tinyiiod *iiod)
...
-void tinyiiod_write_value(struct tinyiiod *iiod, int value)
+/***************************************************************************//**
+ * @brief tinyiiod_write_value
+*******************************************************************************/
+void tinyiiod_write_value(struct tinyiiod *iiod, int32_t value)
 {
 	char buf[16];
 
-	snprintf(buf, sizeof(buf), "%i\n", value);
+	snprintf(buf, sizeof(buf), "%"PRIi32"\n", value);
```
Comment (commodo) [inline tinyiiod.c:152]: the `PRIi32` macro may need to be put into a `compat.h` file and checked whether `#ifndef PRIi32`
i am not sure how standard this is

File: tinyiiod.c [hunk 19]
```diff
@@ -62,160 +74,219 @@ int tinyiiod_read_command(struct tinyiiod *iiod)
...
 {
-	iiod->ops->read(buf, len);
+	return iiod->ops->read(&iiod->instance_id, buf, len);
 }
 
-ssize_t tinyiiod_read_line(struct tinyiiod *iiod, char *buf, size_t len)
+/***************************************************************************//**
+ * @brief tinyiiod_read_line
+*******************************************************************************/
+int32_t tinyiiod_read_line(struct tinyiiod *iiod, char *buf)
```
Comment (mirceacaprioru) [inline tinyiiod.c:100]: Do we need this wrapper anymore? We could just call iiod->ops->read_line(&iiod->instance_id, buf); and just use the iiod object.

**Comment by commodo** [comment]
let's step back a bit here and change how we handle this repo;
this repo, looks like it needs some build integration;

let's start from there

File: tinyiiod.c (hunk 17 above)
Comment (mirceacaprioru) [inline tinyiiod.c:127]: this line is valid here. It separated the variable defines from the code block.

File: tinyiiod.c (hunk 16 above)
Comment (mirceacaprioru) [inline tinyiiod.c:197]: To eliminate this large buffer we need a mechanism to separate the message beforehand and just give chunks of messages where we limit the size (ex: split the message in 128 bytes chunks and send multiple times). This is just an idea not sure of all the implications.

File: tinyiiod.c [hunk 20]
```diff
@@ -62,160 +74,219 @@ int tinyiiod_read_command(struct tinyiiod *iiod)
...
+	char *pbuf; // pointer to memory, where DMA writes, memory free not necessary
 	uint32_t mask;
-	bool print_mask = true;
 
 	ret = iiod->ops->get_mask(device, &mask);
 	if (ret < 0) {
 		tinyiiod_write_value(iiod, ret);
 		return;
 	}
+	// read device
```
Comment (mirceacaprioru) [inline tinyiiod.c:282]: comments should all be consistent with C style like /* Something */.

Comment (commodo) [reply tinyiiod.c:282]: @amiclaus 
can astyle handle this ? i.e. complaining about C style comments ?
i do feel this is something we could automate
